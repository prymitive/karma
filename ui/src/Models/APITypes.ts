export type AlertStateT = "unprocessed" | "active" | "suppressed";

export type LabelsT = { [key: string]: string };

export interface AlertmanagerSilenceMatcherT {
  name: string;
  value: string;
  isRegex: boolean;
  isEqual: boolean;
}

export interface AlertmanagerSilencePayloadT {
  id?: string;
  matchers: AlertmanagerSilenceMatcherT[];
  startsAt: string;
  endsAt: string;
  createdBy: string;
  comment: string;
}

export interface APIAnnotationT {
  name: string;
  value: string;
  visible: boolean;
  isLink: boolean;
  isAction: boolean;
}

export interface APIAlertmanagerStateT {
  fingerprint: string;
  name: string;
  cluster: string;
  state: AlertStateT;
  startsAt: string;
  source: string;
  silencedBy: string[];
  inhibitedBy: string[];
}

export interface APIAlertT {
  id: string;
  annotations: APIAnnotationT[];
  labels: LabelsT;
  startsAt: string;
  state: AlertStateT;
  alertmanager: APIAlertmanagerStateT[];
  receiver: string;
}

export interface StateCountT {
  active: number;
  suppressed: number;
  unprocessed: number;
}

export interface APIAlertGroupT {
  id: string;
  receiver: string;
  labels: LabelsT;
  alerts: APIAlertT[];
  alertmanagerCount: { [key: string]: number };
  stateCount: StateCountT;
  shared: {
    annotations: APIAnnotationT[];
    labels: LabelsT;
    silences: { [cluster: string]: string[] };
  };
}

export interface APISilenceT {
  id: string;
  matchers: AlertmanagerSilenceMatcherT[];
  startsAt: string;
  endsAt: string;
  createdAt: string;
  createdBy: string;
  comment: string;
  ticketID: string;
  ticketURL: string;
}

export interface APIManagedSilenceT {
  alertCount: number;
  cluster: string;
  isExpired: boolean;
  silence: APISilenceT;
}

export interface APIGridT {
  labelName: string;
  labelValue: string;
  alertGroups: APIAlertGroupT[];
  stateCount: StateCountT;
}

export interface APIAlertmanagerUpstreamT {
  name: string;
  cluster: string;
  uri: string;
  publicURI: string;
  readonly: boolean;
  headers: { [key: string]: string };
  corsCredentials: "omit" | "same-origin" | "include";
  error: string;
  version: string;
  clusterMembers: string[];
}

export interface APIFilterT {
  text: string;
  name: string;
  matcher: string;
  value: string;
  hits: number;
  isValid: boolean;
}

export interface APIAlertsResponseUpstreamsCountersT {
  total: number;
  healthy: number;
  failed: number;
}

export type APIAlertsResponseUpstreamsClusterMapT = {
  [clusterName: string]: string[];
};

export type APIAlertsResponseSilenceMapT = {
  [clusterName: string]: { [silenceID: string]: APISilenceT };
};

export interface APIAlertsResponseAuthenticationT {
  enabled: boolean;
  username: string;
}

export interface APILabelColorT {
  brightness: number;
  background: string;
}
export type APIAlertsResponseColorsT = {
  [labelName: string]: { [labelValue: string]: APILabelColorT };
};

export interface APIAlertsResponseUpstreamsT {
  counters: APIAlertsResponseUpstreamsCountersT;
  instances: APIAlertmanagerUpstreamT[];
  clusters: APIAlertsResponseUpstreamsClusterMapT;
}

export interface APILabelCounterValueT {
  value: string;
  raw: string;
  hits: number;
  percent: number;
  offset: number;
}

export interface APILabelCounterT {
  name: string;
  values: APILabelCounterValueT[];
  hits: number;
}

export interface APISettingsT {
  staticColorLabels: string[];
  annotationsDefaultHidden: boolean;
  annotationsHidden: string[];
  annotationsVisible: string[];
  annotationsEnableHTML: boolean;
  sorting: {
    grid: {
      order: string;
      reverse: boolean;
      label: string;
    };
    valueMapping: { [labelName: string]: { [labelValue: string]: number } };
  };
  silenceForm: {
    strip: {
      labels: string[];
    };
  };
  alertAcknowledgement: {
    enabled: boolean;
    durationSeconds: number;
    author: string;
    comment: string;
  };
  historyEnabled: boolean;
}

export interface APIAlertsResponseT {
  status: string;
  error: string;
  timestamp: string;
  version: string;
  upstreams: APIAlertsResponseUpstreamsT;
  silences: APIAlertsResponseSilenceMapT;
  grids: APIGridT[];
  totalAlerts: number;
  colors: APIAlertsResponseColorsT;
  filters: APIFilterT[];
  counters: APILabelCounterT[];
  settings: APISettingsT;
  authentication: APIAlertsResponseAuthenticationT;
  receivers: string[];
}

export interface HistorySampleT {
  timestamp: string;
  value: number;
}

export interface HistoryResponseT {
  error: string;
  samples: HistorySampleT[];
}
