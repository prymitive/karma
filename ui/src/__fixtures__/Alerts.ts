import {
  APIAlertmanagerUpstreamT,
  APIAlertGroupT,
  APIAlertT,
  APISilenceT,
  APIAnnotationT,
  AlertStateT,
  LabelsT,
} from "Models/APITypes";

const MockAnnotation = (
  name: string,
  value: string,
  visible: boolean,
  isLink: boolean,
  isAction: boolean
): APIAnnotationT => ({
  name: name,
  value: value,
  visible: visible,
  isLink: isLink,
  isAction: isAction,
});

const MockAlert = (
  annotations: APIAnnotationT[],
  labels: LabelsT,
  state: AlertStateT
): APIAlertT => ({
  id: Math.random().toString(36),
  annotations: annotations,
  labels: labels,
  startsAt: "2018-08-14T17:36:40.017867056Z",
  state: state,
  alertmanager: [
    {
      fingerprint: "1234567",
      name: "default",
      cluster: "default",
      state: state,
      startsAt: "2018-08-14T17:36:40.017867056Z",
      source: "localhost/graph",
      silencedBy: [],
      inhibitedBy: [],
    },
  ],
  receiver: "by-name",
});

const MockAlertGroup = (
  rootLabels: LabelsT,
  alerts: APIAlertT[],
  sharedAnnotations: APIAnnotationT[],
  sharedLabels: LabelsT,
  sharedSilences: { [cluster: string]: string[] }
): APIAlertGroupT => ({
  receiver: "by-name",
  labels: rootLabels,
  alerts: alerts,
  id: "099c5ca6d1c92f615b13056b935d0c8dee70f18c",
  alertmanagerCount: {
    default: 1,
  },
  stateCount: {
    active: 1,
    suppressed: 0,
    unprocessed: 0,
  },
  shared: {
    annotations: sharedAnnotations,
    labels: sharedLabels,
    silences: sharedSilences,
  },
});

const MockSilence = (): APISilenceT => ({
  comment: "Mocked Silence",
  createdAt: "0001-01-01T00:00:00Z",
  createdBy: "me@example.com",
  startsAt: "2000-01-01T00:00:00Z",
  endsAt: "2000-01-01T01:00:00Z",
  id: "04d37636-2350-4878-b382-e0b50353230f",
  ticketID: "",
  ticketURL: "",
  matchers: [
    { name: "regex", value: "equal", isRegex: true, isEqual: true },
    { name: "regex", value: "notEqual", isRegex: true, isEqual: false },
    { name: "notRegex", value: "equal", isRegex: false, isEqual: true },
    { name: "notRegex", value: "notEqual", isRegex: false, isEqual: false },
  ],
});

const MockAlertmanager = (): APIAlertmanagerUpstreamT => ({
  name: "default",
  cluster: "default",
  uri: "http://localhost",
  publicURI: "http://am.example.com",
  readonly: false,
  headers: {
    Authorization: "Basic foo bar",
  },
  corsCredentials: "include",
  error: "",
  version: "0.17.0",
  clusterMembers: ["default"],
});

export {
  MockAnnotation,
  MockAlert,
  MockAlertGroup,
  MockSilence,
  MockAlertmanager,
};
