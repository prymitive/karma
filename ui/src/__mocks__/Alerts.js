const MockAnnotation = (name, value, visible, isLink) => ({
  name: name,
  value: value,
  visible: visible,
  isLink: isLink,
});

const MockAlert = (annotations, labels, state) => ({
  id: Math.random().toString(36),
  annotations: annotations,
  labels: labels,
  startsAt: "2018-08-14T17:36:40.017867056Z",
  state: state,
  alertmanager: [
    {
      name: "default",
      cluster: "default",
      state: "active",
      startsAt: "2018-08-14T17:36:40.017867056Z",
      source: "localhost/prometheus",
      silencedBy: [],
      inhibitedBy: [],
    },
  ],
  receiver: "by-name",
});

const MockAlertGroup = (
  rootLabels,
  alerts,
  sharedAnnotations,
  sharedLabels,
  sharedSilences
) => ({
  receiver: "by-name",
  labels: rootLabels,
  alerts: alerts,
  id: "099c5ca6d1c92f615b13056b935d0c8dee70f18c",
  alertmanagerCount: {
    default: 1,
  },
  stateCount: {
    active: 1,
    suppressed: 0,
    unprocessed: 0,
  },
  shared: {
    annotations: sharedAnnotations,
    labels: sharedLabels,
    silences: sharedSilences,
  },
});

const MockSilence = () => ({
  comment: "Mocked Silence",
  createdAt: "0001-01-01T00:00:00Z",
  createdBy: "me@example.com",
  startsAt: "2000-01-01T00:00:00Z",
  endsAt: "2000-01-01T01:00:00Z",
  id: "04d37636-2350-4878-b382-e0b50353230f",
  ticketID: "",
  ticketURL: "",
  matchers: [
    { name: "foo", value: "bar", isRegex: false },
    { name: "baz", value: "regex", isRegex: true },
  ],
});

const MockAlertmanager = () => ({
  name: "default",
  cluster: "default",
  uri: "http://localhost",
  publicURI: "http://am.example.com",
  readonly: false,
  headers: {
    Authorization: "Basic foo bar",
  },
  corsCredentials: "include",
  error: "",
  version: "0.17.0",
  clusterMembers: ["default"],
});

export {
  MockAnnotation,
  MockAlert,
  MockAlertGroup,
  MockSilence,
  MockAlertmanager,
};
