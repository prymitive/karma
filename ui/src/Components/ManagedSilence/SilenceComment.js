import React from "react";
import PropTypes from "prop-types";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faExternalLinkAlt } from "@fortawesome/free-solid-svg-icons/faExternalLinkAlt";
import { faBellSlash } from "@fortawesome/free-solid-svg-icons/faBellSlash";

import { APISilence } from "Models/API";
import { AlertStore } from "Stores/AlertStore";
import { FilteringCounterBadge } from "Components/Labels/FilteringCounterBadge";
import { ToggleIcon } from "Components/ToggleIcon";
import { SilenceProgress } from "./SilenceProgress";

const SilenceComment = ({
  cluster,
  silence,
  alertCount,
  alertCountAlwaysVisible,
  collapsed,
  collapseToggle,
  afterUpdate,
  alertStore,
}) => {
  const comment = silence.ticketURL ? (
    <a
      href={silence.ticketURL}
      target="_blank"
      rel="noopener noreferrer"
      className={
        collapsed
          ? "components-managed-silence-comment mw-100 text-truncate d-block"
          : "components-managed-silence-comment"
      }
    >
      <FontAwesomeIcon className="mr-2" icon={faExternalLinkAlt} />
      {silence.comment}
    </a>
  ) : (
    silence.comment
  );

  return (
    <React.Fragment>
      <div className="d-flex flex-row">
        <div className="flex-shrink-0 flex-grow-0">
          <FontAwesomeIcon
            icon={faBellSlash}
            className="components-managed-silence-icon text-muted"
          />
        </div>
        <div className="mx-2 flex-shrink-1 flex-grow-1 mw-1p">
          <div
            className={`font-italic components-managed-silence-comment ${
              collapsed ? "text-truncate overflow-hidden" : ""
            }`}
          >
            {comment}
          </div>
          <div className="components-managed-silence-cite mt-1">
            <span className="text-muted mr-2 font-italic">
              &mdash; {silence.createdBy}
            </span>
            {collapsed &&
            Object.keys(alertStore.data.upstreams.clusters).length > 1 ? (
              <span className="badge badge-secondary mx-1 align-text-bottom p-1">
                {cluster}
              </span>
            ) : null}
            {collapsed ? <SilenceProgress silence={silence} /> : null}
          </div>
        </div>
        <div className="flex-shrink-0 flex-grow-0">
          <div className="d-flex flex-column flex-sm-row justify-content-between align-items-center">
            <FilteringCounterBadge
              alertStore={alertStore}
              name="@silence_id"
              value={silence.id}
              counter={alertCount}
              themed={false}
              alwaysVisible={alertCountAlwaysVisible}
              defaultColor="primary"
            />
            <ToggleIcon
              isOpen={!collapsed}
              className={`components-managed-silence-icon ${
                alertCount && alertCountAlwaysVisible && "my-sm-auto mt-2 mb-0"
              } ml-sm-2 ml-auto mr-sm-0 mr-1 text-muted cursor-pointer`}
              onClick={collapseToggle}
            />
          </div>
        </div>
      </div>
    </React.Fragment>
  );
};
SilenceComment.propTypes = {
  cluster: PropTypes.string.isRequired,
  silence: APISilence.isRequired,
  alertCount: PropTypes.number.isRequired,
  collapsed: PropTypes.bool.isRequired,
  collapseToggle: PropTypes.func.isRequired,
  alertStore: PropTypes.instanceOf(AlertStore).isRequired,
};

export { SilenceComment };
