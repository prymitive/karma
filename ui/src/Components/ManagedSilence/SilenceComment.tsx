import { FC } from "react";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faExternalLinkAlt } from "@fortawesome/free-solid-svg-icons/faExternalLinkAlt";
import { faBellSlash } from "@fortawesome/free-solid-svg-icons/faBellSlash";

import { APISilenceT } from "Models/APITypes";
import { AlertStore } from "Stores/AlertStore";
import FilteringCounterBadge from "Components/Labels/FilteringCounterBadge";
import { ToggleIcon } from "Components/ToggleIcon";
import { SilenceProgress } from "./SilenceProgress";

const SilenceComment: FC<{
  cluster: string;
  silence: APISilenceT;
  alertCount: number;
  collapsed: boolean;
  collapseToggle: () => void;
  alertStore: AlertStore;
  alertCountAlwaysVisible?: boolean;
}> = ({
  cluster,
  silence,
  alertCount,
  alertCountAlwaysVisible,
  collapsed,
  collapseToggle,
  alertStore,
}) => {
  const comment = silence.comment.split(" ").map((w, i) =>
    silence.ticketURL && w === silence.ticketID ? (
      <a
        key={i}
        href={silence.ticketURL}
        target="_blank"
        rel="noopener noreferrer"
      >
        <FontAwesomeIcon className="me-2" icon={faExternalLinkAlt} />
        {silence.ticketID}
      </a>
    ) : (
      " " + w
    )
  );

  return (
    <>
      <div className="d-flex flex-row">
        <div className="flex-shrink-0 flex-grow-0">
          <FontAwesomeIcon
            icon={faBellSlash}
            className="components-managed-silence-icon text-muted"
          />
        </div>
        <div className="mx-2 flex-shrink-1 flex-grow-1 mw-1p">
          <div
            className={`font-italic components-managed-silence-comment ${
              collapsed ? "text-truncate overflow-hidden" : ""
            }`}
          >
            {comment.map((w) => w)}
          </div>
          <div className="components-managed-silence-cite mt-1">
            <span className="text-muted me-2 font-italic">
              &mdash; {silence.createdBy}
            </span>
            {collapsed &&
            Object.keys(alertStore.data.upstreams.clusters).length > 1 ? (
              <span className="badge bg-secondary mx-1 align-text-bottom p-1">
                {cluster}
              </span>
            ) : null}
            {collapsed ? <SilenceProgress silence={silence} /> : null}
          </div>
        </div>
        <div className="flex-shrink-0 flex-grow-0">
          <div className="d-flex flex-column flex-sm-row justify-content-between align-items-center">
            <FilteringCounterBadge
              alertStore={alertStore}
              name="@silence_id"
              value={silence.id}
              counter={alertCount}
              themed={false}
              alwaysVisible={alertCountAlwaysVisible}
              defaultColor="primary"
              isAppend={false}
            />
            <span className="badge components-label with-click">
              <ToggleIcon
                isOpen={!collapsed}
                className="components-managed-silence-icon m-auto text-muted cursor-pointer"
                onClick={collapseToggle}
              />
            </span>
          </div>
        </div>
      </div>
    </>
  );
};

export { SilenceComment };
