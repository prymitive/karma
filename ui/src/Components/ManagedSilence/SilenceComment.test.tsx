import { mount } from "enzyme";

import { advanceTo, clear } from "jest-date-mock";

import toDiffableHtml from "diffable-html";

import { MockSilence } from "__fixtures__/Alerts";
import { APISilenceT } from "Models/APITypes";
import { AlertStore } from "Stores/AlertStore";
import { SilenceComment } from "./SilenceComment";

let silence: APISilenceT;
let alertStore: AlertStore;

beforeEach(() => {
  advanceTo(new Date(Date.UTC(2000, 0, 1, 0, 30, 0)));

  silence = MockSilence();
  alertStore = new AlertStore([]);
});

afterEach(() => {
  jest.restoreAllMocks();
  clear();
});

const CollapseMock = jest.fn();

const MountedSilenceComment = (collapsed: boolean, cluster?: string) => {
  return mount(
    <SilenceComment
      alertStore={alertStore}
      alertCount={123}
      cluster={cluster || "default"}
      silence={silence}
      collapsed={collapsed}
      collapseToggle={CollapseMock}
    />
  );
};

const MockMultipleClusters = () => {
  alertStore.data.setUpstreams({
    counters: { total: 2, healthy: 2, failed: 0 },
    clusters: { ha: ["ha1", "ha2"], single: ["single"] },
    instances: [
      {
        name: "ha1",
        uri: "http://ha1.example.com",
        publicURI: "http://ha1.example.com",
        readonly: false,
        headers: {},
        corsCredentials: "include",
        error: "",
        version: "0.17.0",
        cluster: "ha",
        clusterMembers: ["ha1", "ha2"],
      },
      {
        name: "ha2",
        uri: "http://ha2.example.com",
        publicURI: "http://ha2.example.com",
        readonly: false,
        headers: {},
        corsCredentials: "include",
        error: "",
        version: "0.17.0",
        cluster: "ha",
        clusterMembers: ["ha1", "ha2"],
      },
      {
        name: "single",
        uri: "http://single.example.com",
        publicURI: "http://single.example.com",
        readonly: false,
        headers: {},
        corsCredentials: "include",
        error: "",
        version: "0.17.0",
        cluster: "single",
        clusterMembers: ["single"],
      },
    ],
  });
};

describe("<SilenceComment />", () => {
  it("Matches snapshot when collapsed", () => {
    const tree = MountedSilenceComment(true);
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
  });

  it("Matches snapshot when expanded", () => {
    const tree = MountedSilenceComment(false);
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
  });

  it("Matches snapshot when collapsed and multiple clusters are present", () => {
    MockMultipleClusters();
    const tree = MountedSilenceComment(true, "ha");
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
  });

  it("Matches snapshot when expanded and multiple clusters are present", () => {
    MockMultipleClusters();
    const tree = MountedSilenceComment(false, "ha");
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
  });

  it("Renders a JIRA link if present", () => {
    silence.ticketURL = "http://localhost/1234";
    silence.ticketID = "1234";
    silence.comment = "Ticket id 1234 and also 1234";
    const tree = MountedSilenceComment(true);
    expect(tree.find("a[href='http://localhost/1234']")).toHaveLength(2);
  });

  it("Renders a JIRA link if present and comment is expanded", () => {
    silence.ticketURL = "http://localhost/1234";
    silence.ticketID = "1234";
    silence.comment = "Ticket id 1234";
    const tree = MountedSilenceComment(false);
    expect(tree.find("a[href='http://localhost/1234']")).toHaveLength(1);
  });

  it("collapseToggle is called when collapse icon is clicked", () => {
    const tree = MountedSilenceComment(true);
    const collapse = tree.find("svg.fa-chevron-down");
    collapse.simulate("click");
    expect(CollapseMock).toHaveBeenCalled();
  });

  it("Doesn't render cluster badges when collapsed and only a single cluster is present", () => {
    const tree = MountedSilenceComment(true);
    const ams = tree.find("span.badge.bg-secondary");
    expect(ams).toHaveLength(0);
  });

  it("Doesn't render cluster badges when expanded and only a single cluster is present", () => {
    const tree = MountedSilenceComment(false);
    const ams = tree.find("span.badge.bg-secondary");
    expect(ams).toHaveLength(0);
  });

  it("Renders cluster badge when collapsed and multiple clusters are present", () => {
    MockMultipleClusters();
    const tree = MountedSilenceComment(true, "single");
    const ams = tree.find("span.badge.bg-secondary");
    expect(ams).toHaveLength(1);
    expect(toDiffableHtml(ams.at(0).html())).toMatch(/single/);
  });

  it("Doesn't render cluster badge when expanded and multiple clusters are present", () => {
    MockMultipleClusters();
    const tree = MountedSilenceComment(false, "single");
    const ams = tree.find("span.badge.bg-secondary");
    expect(ams).toHaveLength(0);
  });
});
