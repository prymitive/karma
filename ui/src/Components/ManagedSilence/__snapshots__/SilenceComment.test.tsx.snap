// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SilenceComment /> Matches snapshot when collapsed 1`] = `
"
<div class="d-flex flex-row">
  <div class="flex-shrink-0 flex-grow-0">
    <svg
      aria-hidden="true"
      focusable="false"
      data-prefix="fas"
      data-icon="bell-slash"
      class="svg-inline--fa fa-bell-slash text-muted"
      role="img"
      xmlns="http://www.w3.org/2000/svg"
      viewbox="0 0 640 512"
    >
      <path
        fill="currentColor"
        d="M38.8 5.1C28.4-3.1 13.3-1.2 5.1 9.2S-1.2 34.7 9.2 42.9l592 464c10.4 8.2 25.5 6.3 33.7-4.1s6.3-25.5-4.1-33.7l-90.2-70.7c.2-.4 .4-.9 .6-1.3c5.2-11.5 3.1-25-5.3-34.4l-7.4-8.3C497.3 319.2 480 273.9 480 226.8V208c0-77.4-55-142-128-156.8V32c0-17.7-14.3-32-32-32s-32 14.3-32 32V51.2c-42.6 8.6-79 34.2-102 69.3L38.8 5.1zM406.2 416L160 222.1v4.8c0 47-17.3 92.4-48.5 127.6l-7.4 8.3c-8.4 9.4-10.4 22.9-5.3 34.4S115.4 416 128 416H406.2zm-40.9 77.3c12-12 18.7-28.3 18.7-45.3H320 256c0 17 6.7 33.3 18.7 45.3s28.3 18.7 45.3 18.7s33.3-6.7 45.3-18.7z"
      >
      </path>
    </svg>
  </div>
  <div class="mx-2 flex-shrink-1 flex-grow-1 mw-1p">
    <div class="components-managed-silence-comment text-truncate overflow-hidden">
      Mocked Silence
    </div>
    <div class="components-managed-silence-cite mt-1 d-flex flex-row">
      <span class="text-muted text-truncate overflow-hidden me-2">
        — me@example.com
      </span>
      <div class="d-flex flex-row justify-content-end flex-grow-1">
        <span class="badge bg-light components-label">
          Expires in 30 minutes
        </span>
      </div>
    </div>
  </div>
  <div class="flex-shrink-0 flex-grow-0">
    <div class="d-flex flex-column flex-sm-row justify-content-between align-items-center">
      <div class=" tooltip-trigger">
        <span class="bg-primary rounded-pill components-label-with-hover components-label badge">
          123
        </span>
      </div>
      <span class="badge components-label with-click">
        <svg
          aria-hidden="true"
          focusable="false"
          data-prefix="fas"
          data-icon="chevron-down"
          class="svg-inline--fa fa-chevron-down fa-rotate-180 m-auto text-muted cursor-pointer"
          role="img"
          xmlns="http://www.w3.org/2000/svg"
          viewbox="0 0 512 512"
          style="transition: transform 0.25s ease-in-out;"
        >
          <path
            fill="currentColor"
            d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
          >
          </path>
        </svg>
      </span>
    </div>
  </div>
</div>
"
`;

exports[`<SilenceComment /> Matches snapshot when collapsed and multiple clusters are present 1`] = `
"
<div class="d-flex flex-row">
  <div class="flex-shrink-0 flex-grow-0">
    <svg
      aria-hidden="true"
      focusable="false"
      data-prefix="fas"
      data-icon="bell-slash"
      class="svg-inline--fa fa-bell-slash text-muted"
      role="img"
      xmlns="http://www.w3.org/2000/svg"
      viewbox="0 0 640 512"
    >
      <path
        fill="currentColor"
        d="M38.8 5.1C28.4-3.1 13.3-1.2 5.1 9.2S-1.2 34.7 9.2 42.9l592 464c10.4 8.2 25.5 6.3 33.7-4.1s6.3-25.5-4.1-33.7l-90.2-70.7c.2-.4 .4-.9 .6-1.3c5.2-11.5 3.1-25-5.3-34.4l-7.4-8.3C497.3 319.2 480 273.9 480 226.8V208c0-77.4-55-142-128-156.8V32c0-17.7-14.3-32-32-32s-32 14.3-32 32V51.2c-42.6 8.6-79 34.2-102 69.3L38.8 5.1zM406.2 416L160 222.1v4.8c0 47-17.3 92.4-48.5 127.6l-7.4 8.3c-8.4 9.4-10.4 22.9-5.3 34.4S115.4 416 128 416H406.2zm-40.9 77.3c12-12 18.7-28.3 18.7-45.3H320 256c0 17 6.7 33.3 18.7 45.3s28.3 18.7 45.3 18.7s33.3-6.7 45.3-18.7z"
      >
      </path>
    </svg>
  </div>
  <div class="mx-2 flex-shrink-1 flex-grow-1 mw-1p">
    <div class="components-managed-silence-comment text-truncate overflow-hidden">
      Mocked Silence
    </div>
    <div class="components-managed-silence-cite mt-1 d-flex flex-row">
      <span class="text-muted text-truncate overflow-hidden me-2">
        — me@example.com
      </span>
      <div class="d-flex flex-row justify-content-end flex-grow-1">
        <span class="badge bg-light components-label">
          Expires in 30 minutes
        </span>
        <span class="badge bg-secondary mx-1 components-label">
          ha
        </span>
      </div>
    </div>
  </div>
  <div class="flex-shrink-0 flex-grow-0">
    <div class="d-flex flex-column flex-sm-row justify-content-between align-items-center">
      <div class=" tooltip-trigger">
        <span class="bg-primary rounded-pill components-label-with-hover components-label badge">
          123
        </span>
      </div>
      <span class="badge components-label with-click">
        <svg
          aria-hidden="true"
          focusable="false"
          data-prefix="fas"
          data-icon="chevron-down"
          class="svg-inline--fa fa-chevron-down fa-rotate-180 m-auto text-muted cursor-pointer"
          role="img"
          xmlns="http://www.w3.org/2000/svg"
          viewbox="0 0 512 512"
          style="transition: transform 0.25s ease-in-out;"
        >
          <path
            fill="currentColor"
            d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
          >
          </path>
        </svg>
      </span>
    </div>
  </div>
</div>
"
`;

exports[`<SilenceComment /> Matches snapshot when expanded 1`] = `
"
<div class="d-flex flex-row">
  <div class="flex-shrink-0 flex-grow-0">
    <svg
      aria-hidden="true"
      focusable="false"
      data-prefix="fas"
      data-icon="bell-slash"
      class="svg-inline--fa fa-bell-slash text-muted"
      role="img"
      xmlns="http://www.w3.org/2000/svg"
      viewbox="0 0 640 512"
    >
      <path
        fill="currentColor"
        d="M38.8 5.1C28.4-3.1 13.3-1.2 5.1 9.2S-1.2 34.7 9.2 42.9l592 464c10.4 8.2 25.5 6.3 33.7-4.1s6.3-25.5-4.1-33.7l-90.2-70.7c.2-.4 .4-.9 .6-1.3c5.2-11.5 3.1-25-5.3-34.4l-7.4-8.3C497.3 319.2 480 273.9 480 226.8V208c0-77.4-55-142-128-156.8V32c0-17.7-14.3-32-32-32s-32 14.3-32 32V51.2c-42.6 8.6-79 34.2-102 69.3L38.8 5.1zM406.2 416L160 222.1v4.8c0 47-17.3 92.4-48.5 127.6l-7.4 8.3c-8.4 9.4-10.4 22.9-5.3 34.4S115.4 416 128 416H406.2zm-40.9 77.3c12-12 18.7-28.3 18.7-45.3H320 256c0 17 6.7 33.3 18.7 45.3s28.3 18.7 45.3 18.7s33.3-6.7 45.3-18.7z"
      >
      </path>
    </svg>
  </div>
  <div class="mx-2 flex-shrink-1 flex-grow-1 mw-1p">
    <div class="components-managed-silence-comment ">
      Mocked Silence
    </div>
    <div class="components-managed-silence-cite mt-1 d-flex flex-row">
      <span class="text-muted text-truncate overflow-hidden ">
        — me@example.com
      </span>
    </div>
  </div>
  <div class="flex-shrink-0 flex-grow-0">
    <div class="d-flex flex-column flex-sm-row justify-content-between align-items-center">
      <div class=" tooltip-trigger">
        <span class="bg-primary rounded-pill components-label-with-hover components-label badge">
          123
        </span>
      </div>
      <span class="badge components-label with-click">
        <svg
          aria-hidden="true"
          focusable="false"
          data-prefix="fas"
          data-icon="chevron-down"
          class="svg-inline--fa fa-chevron-down m-auto text-muted cursor-pointer"
          role="img"
          xmlns="http://www.w3.org/2000/svg"
          viewbox="0 0 512 512"
          style="transition: transform 0.25s ease-in-out;"
        >
          <path
            fill="currentColor"
            d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
          >
          </path>
        </svg>
      </span>
    </div>
  </div>
</div>
"
`;

exports[`<SilenceComment /> Matches snapshot when expanded and multiple clusters are present 1`] = `
"
<div class="d-flex flex-row">
  <div class="flex-shrink-0 flex-grow-0">
    <svg
      aria-hidden="true"
      focusable="false"
      data-prefix="fas"
      data-icon="bell-slash"
      class="svg-inline--fa fa-bell-slash text-muted"
      role="img"
      xmlns="http://www.w3.org/2000/svg"
      viewbox="0 0 640 512"
    >
      <path
        fill="currentColor"
        d="M38.8 5.1C28.4-3.1 13.3-1.2 5.1 9.2S-1.2 34.7 9.2 42.9l592 464c10.4 8.2 25.5 6.3 33.7-4.1s6.3-25.5-4.1-33.7l-90.2-70.7c.2-.4 .4-.9 .6-1.3c5.2-11.5 3.1-25-5.3-34.4l-7.4-8.3C497.3 319.2 480 273.9 480 226.8V208c0-77.4-55-142-128-156.8V32c0-17.7-14.3-32-32-32s-32 14.3-32 32V51.2c-42.6 8.6-79 34.2-102 69.3L38.8 5.1zM406.2 416L160 222.1v4.8c0 47-17.3 92.4-48.5 127.6l-7.4 8.3c-8.4 9.4-10.4 22.9-5.3 34.4S115.4 416 128 416H406.2zm-40.9 77.3c12-12 18.7-28.3 18.7-45.3H320 256c0 17 6.7 33.3 18.7 45.3s28.3 18.7 45.3 18.7s33.3-6.7 45.3-18.7z"
      >
      </path>
    </svg>
  </div>
  <div class="mx-2 flex-shrink-1 flex-grow-1 mw-1p">
    <div class="components-managed-silence-comment ">
      Mocked Silence
    </div>
    <div class="components-managed-silence-cite mt-1 d-flex flex-row">
      <span class="text-muted text-truncate overflow-hidden ">
        — me@example.com
      </span>
    </div>
  </div>
  <div class="flex-shrink-0 flex-grow-0">
    <div class="d-flex flex-column flex-sm-row justify-content-between align-items-center">
      <div class=" tooltip-trigger">
        <span class="bg-primary rounded-pill components-label-with-hover components-label badge">
          123
        </span>
      </div>
      <span class="badge components-label with-click">
        <svg
          aria-hidden="true"
          focusable="false"
          data-prefix="fas"
          data-icon="chevron-down"
          class="svg-inline--fa fa-chevron-down m-auto text-muted cursor-pointer"
          role="img"
          xmlns="http://www.w3.org/2000/svg"
          viewbox="0 0 512 512"
          style="transition: transform 0.25s ease-in-out;"
        >
          <path
            fill="currentColor"
            d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
          >
          </path>
        </svg>
      </span>
    </div>
  </div>
</div>
"
`;
